libraries/AP_L1_Control/AP_L1_Control.cpp

```diff
@@ -442,18 +443,39 @@ void AP_L1_Control::update_loiter(const Location &center_WP, float radius, int8_
     // Perform switchover between 'capture' and 'circle' modes at the
     // point where the commands cross over to achieve a seamless transfer
     // Only fly 'capture' mode if outside the circle
+    const uint32_t now_ms = AP_HAL::millis();
     if (xtrackErrCirc > 0.0f && loiter_direction * latAccDemCap < loiter_direction * latAccDemCirc) {
         _latAccDem = latAccDemCap;
-        _WPcircle = false;
+
+        /*
+          if we were previously on the circle and the target has not
+          changed then keep _WPcircle true. This prevents
+          reached_loiter_target() from going false due to a gust of
+          wind or an unachievable loiter radius
+         */
+        if (_WPcircle &&
+            last_reached_loiter_target_ms != 0 &&
+            now_ms - last_reached_loiter_target_ms < 200U &&
+            center_WP.same_loc_as(last_center_WP)) {
```

Comment:
Should we also check the radius is the same? Going from a large radius to a smaller one would previously result in `_WPcircle` false. Now it will not. Small to large is OK because we would be inside the radius already.

Reply:
I didn't add radius as it varies with height, but I've added it now using the unscaled radius from the entry to the function. Also added loiter direction